### ISSUE TITLE
Async methods are blocking execution

### ISSUE THREAD

**Issue by thiromi** [StackExchange/dapper-dot-net#531]
Aparently, `*Async` methods are not so "async", as its name suggests.
I have the following code:
```C#
using Dapper;
using MySql.Data.MySqlClient;
using System;
using System.Data.Common;
using System.Diagnostics;

namespace ConsoleApplication
{
    public class Program
    {
        public class Something
        {
            public string Name { get; set; }
            public string Email { get; set; }
        }
        public static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
            var connectionString = String.Format(
                "Server={0};Port=3306;Database={1};Uid={2};Pwd={3};",
                args[0],
                args[1],
                args[2],
                args[3]
            );

            using (DbConnection connection = new MySqlConnection(connectionString))
            {
                Stopwatch profiler = new Stopwatch();
                Console.WriteLine("Making first task");
                profiler.Start();
                var task1 = connection.QueryFirstAsync<Something>("SELECT * from (SELECT 'John Doe' as Name, 'john@doe.com' as Email) data left join (SELECT SLEEP(3)) timer on 1=1;");
                Console.WriteLine("Terminated first task, going to second task. elapsed {0}ms", profiler.ElapsedMilliseconds.ToString());
                profiler.Restart();
                var task2 = connection.QueryFirstAsync<Something>("SELECT * from (SELECT 'Mary Doe' as Name, 'mary@doe.com' as Email) data left join (SELECT SLEEP(3)) timer on 1=1;");
                Console.WriteLine("Terminated second task, getting results. elapsed {0}ms", profiler.ElapsedMilliseconds.ToString());

                task1.Wait();
                task2.Wait();

                var something1 = task1.Result;
                var something2 = task2.Result;

                Console.WriteLine(something1.Name);
                Console.WriteLine(something2.Name);

                Console.WriteLine("End");
            }
        }
    }
}
```

running this code lead me to the following output:

```
Hello World!
Making first task
Terminated first task, going to second task. elapsed 3338ms
Terminated second task, getting results. elapsed 3009ms
John Doe
Mary Doe
End
```

What I expected to see was:

```
Terminated first task, going to second task. elapsed <way less than 3s>
Terminated second task, getting results. elapsed <way less than 3s>
```

This means that all `*Async` methods blocks the execution until DBMS returns the data to the application.

Could you please unblock the execution when fetching data from DBMS?

I am using Dapper 1.50.0-rc2b with ASP.Net 5 rc1 on Mac OSX

**Comment by jamesholwell** [comment]
If I recall, this is an issue with the MySQL connector, I'll try and dig out the previous thread in a bit.

**Comment by mgravell** [comment]
As already noted: whether this is implemented as true async vs
async-over-sync vs whatever else is largely down to the ADO.NET provider.
Note: you probably shouldn't be calling Wait, either - await is preferable
in virtually all scenarios (but I acknowledge that this is problematic for
a console exe Main method)

**Comment by NickCraver** [comment]
To add source links here. Due to history and these methods being added later, `DbDataReader` implements `ReadAsync()` as synchronous so that it's *available*, but nothing about it is `async`. [You can see the code here](https://github.com/Microsoft/referencesource/blob/master/System.Data/System/Data/Common/DbDataReader.cs#L246). Many other async bits added to the core ADO.NET types are similarly implemented. The way this works with providers is it's up to them to implement async by overriding that virtual method.

Unfortunately, the MySQL .NET Connector is pretty neglected, it does not override `ReadAsync()` at all. [It only implements `Read()`](https://github.com/mysql/mysql-connector-net/blob/6.9/Source/MySql.Data/datareader.cs#L908).

We'd love to help here, but there's absolutely nothing Dapper can do. Dapper is properly awaiting and such here, the synchronous piece in play is outside of our control. [There is an open issue to fix this on Oracle's side](https://bugs.mysql.com/bug.php?id=70111), but to be honest...they don't seem to care, which is terrible for everyone.

**Comment by thiromi** [comment]
Oh, OK then. I'll do it in a separate task and wait until they fix this issue. Thanks a lot for your clarification!
